contribute/design/1.x/upgrade/volume-pools-upgrade.md

```diff
@@ -0,0 +1,499 @@
+  watching the BulkUpgrade can then launch individual UpgradeTasks. 
+
+   
+- How does this design compared to the `kubectl` based upgrade
+  introduced for upgrading from [0.8.2 to 0.9](https://github.com/openebs/openebs/tree/master/k8s/upgrades/0.8.2-0.9.0). 
+
+  The current design proposed in this document builds on top of the
+  0.8.2 to 0.9 design, by improving on usability and agility of the 
+  upgrade code development. 
+
+  The implementation made use of the CAS Template feature and comprised 
+  of the following: 
+  - Each type of upgrade - like jiva volume upgrade, cstor pool upgrade, 
+    were converted from shell to an CAS Template, with a series of  
+    upgrade tasks (aka runtasks) to be executed. 
+  - A custom resource called upgraderesults was defined to capture
+    the results while executing the run tasks.
+  - An upgrade agent, that can read the upgrade CAS Template and 
+    execute the run tasks on a given volume or pool.
+  - Administrator will have to create a Kubernetes Job out of the
+    upgrade agent - by passing the upgrade cas template and the object
+    details via a config map. 
+
+  While the above approach was good enough to be automated via `kubectl`, 
+  the steps that can be executed as part of upgrade were limited
+  by the features or constructs available within the CAST/RunTasks. 
+
+  Another short-coming of the CAST/RunTasks is the lack of developer
+  friendly constructs for rapidly writing new CAST/RunTasks. This was
+  severely impacting the time taken to push out new releases. 
+
+### High Level Design
+
+#### Upgrade Job Example
+
+Here is an example of Kubernetes Job spec for upgrading the jiva volume.
+```
+#This is an example YAML for upgrading jiva volume. 
+#Some of the values below needs to be changed to
+#match your openebs installation. The fields are
+#indicated with VERIFY
+---
+apiVersion: batch/v1
+kind: Job
+metadata:
+  #VERIFY that you have provided a unique name for this upgrade job.
+  #The name can be any valid K8s string for name. This example uses
+  #the following convention: jiva-vol-<flattened-from-to-versions>-<pv-name>
+  name: jiva-vol-100110-pvc-713e3bb6-afd2-11e9-8e79-42010a800065
+  #VERIFY the value of namespace is same as the namespace where openebs components
+  # are installed. You can verify using the command:
+  # `kubectl get pods -n <openebs-namespace> -l openebs.io/component-name=maya-apiserver`
+  # The above command should return status of the openebs-apiserver.
+  namespace: openebs
+spec:
+  backoffLimit: 4
+  template:
+    spec:
+      #VERIFY the value of serviceAccountName is pointing to service account
+      # created within openebs namespace. Use the non-default account.
+      # by running `kubectl get sa -n <openebs-namespace>`
+      serviceAccountName: openebs-maya-operator
+      containers:
+      - name:  upgrade
+        args: 
+        - "jiva-volume"
+        - "--from-version=1.0.0"
+        - "--to-version=1.1.0-RC2"
```

Comment:
Can we move the args to ENV. This helps the higher order components to build this Job template in a easier way.

In other words, can we suggest use of ENV variables vs. flag. However this depends entirely on the logic written inside the image.

Reply:
The current design expects the command line variables and it is a bit clear to understand than ENV. As we add higher level operators UpgradeTask CR will be a single parameters passed to the UpgradeJob.